src/embeddedTraceViewer.ts

```diff
@@ -0,0 +1,261 @@
+import { DisposableBase } from './disposableBase';
+import { PlaywrightTestServer } from './playwrightTestServer';
+import { TraceViewer } from './traceViewer';
+
+export class EmbeddedTraceViewer implements TraceViewer {
+  readonly vscode: vscodeTypes.VSCode;
+  readonly extensionUri: vscodeTypes.Uri;
+  private _currentFile?: string;
+  private _traceViewerPanelPromise?: Promise<EmbeddedTraceViewerPanel | undefined>;
+  private _config: TestConfig;
+  private _testServer: PlaywrightTestServer;
+
+  constructor(vscode: vscodeTypes.VSCode, extensionUri: vscodeTypes.Uri, config: TestConfig, testServer: PlaywrightTestServer) {
+    this.vscode = vscode;
+    this.extensionUri = extensionUri;
+    this._config = config;
+    this._testServer = testServer;
+  }
+
+  currentFile() {
+    return this._currentFile;
+  }
+
+  async willRunTests() {
+    await this._startIfNeeded();
+  }
+
+  async open(file?: string) {
+    this._currentFile = file;
+    if (!file && !this._traceViewerPanelPromise)
+      return;
+    const traceViewerPanel = await this._startIfNeeded();
+    traceViewerPanel?.loadTraceRequested(file);
+  }
+
+  close() {
+    this._traceViewerPanelPromise?.then(panel => panel?.dispose()).catch(() => {});
+    this._traceViewerPanelPromise = undefined;
+    this._currentFile = undefined;
+  }
+
+  private async _startIfNeeded() {
+    if (!this._traceViewerPanelPromise)
+      this._traceViewerPanelPromise = this._createTraceViewerPanel();
+    return await this._traceViewerPanelPromise;
+  }
+
+  private async _createTraceViewerPanel() {
+    const serverUrlPrefix = await this._testServer.ensureStartedForTraceViewer();
+    if (!serverUrlPrefix)
+      return;
+    return new EmbeddedTraceViewerPanel(this, serverUrlPrefix);
+  }
+
+  async infoForTest() {
+    const traceViewerPanel = await this._traceViewerPanelPromise;
+    return {
+      type: 'embedded',
+      serverUrlPrefix: traceViewerPanel?.serverUrlPrefix,
+      testConfigFile: this._config.configFile,
+      traceFile: this._currentFile,
+    };
+  }
+}
+
+class EmbeddedTraceViewerPanel extends DisposableBase {
+
+  public static readonly viewType = 'playwright.traceviewer.view';
+
+  private _vscode: vscodeTypes.VSCode;
+  private _extensionUri: vscodeTypes.Uri;
+  private _webviewPanel: vscodeTypes.WebviewPanel;
+  readonly serverUrlPrefix: string;
+  private _isVisible: boolean = false;
+  private _viewColumn: vscodeTypes.ViewColumn | undefined;
+  private _traceUrl?: string;
+  private _traceLoadRequestedTimeout?: NodeJS.Timeout;
+
+  constructor(
+    embeddedTestViewer: EmbeddedTraceViewer,
+    serverUrlPrefix: string
+  ) {
+    super();
+    this._vscode = embeddedTestViewer.vscode;
+    this._extensionUri = embeddedTestViewer.extensionUri;
+    this.serverUrlPrefix = serverUrlPrefix;
+    this._isVisible = false;
+    this._webviewPanel = this._vscode.window.createWebviewPanel(EmbeddedTraceViewerPanel.viewType, 'Trace Viewer', {
```

Comment:
Since we are not implementing an editor of the selected document, we should not use this API to show trace viewer. We should instead show it as a sidebar or a drawer panel and encourage user to drag it to the right. Try doing it with the Playwright settings and you'll see what I mean.

Reply:
Do you mean this?

![image](https://github.com/user-attachments/assets/43d6fd0e-09ee-4c15-a077-99d1265d4d8a)

I honestly don't really like that approach. I'd rather go with editors. I already started to implement it but I left it on hold while this PR was under review.

For reference, here's a screenshot of the current status, with two trace editors side by side:

![image](https://github.com/user-attachments/assets/2e1a5126-cab2-4d48-affe-6af1c489eea5)

I did use the `vscode.window.createWebviewPanel` API because it's the one that offers a similar user experience to editors.

Reply:
Traces side by side is a good example of why you should not be using the editor pattern. It is unclear what controls what. If trace viewer was opened for a trace file, it would be a different story.

Reply:
I created URIs so that they got proper breadcrumbs.

Meanwhle, I was able to implement it with webviews. Here's a screenshot of it:

![image](https://github.com/user-attachments/assets/48703bd8-7cc4-4262-9435-bf29afe80e47)

, I'll clean the code a bit and create a new PR for that when I have some spare time.